platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/OutboundConnectionCreator.java

```diff
@@ -131,23 +138,4 @@ public Connection createConnection(final NodeId otherId) {
 
         return NotConnectedConnection.getSingleton();
     }
-
-    /**
-     * Find the best way to connect <code>from</code> address <code>to</code> address
-     *
-     * @param from the address that needs to connect
-     * @param to   the address to connect to
-     * @return the IP address to connect to
-     */
-    private String getConnectHostname(final Address from, final Address to) {
```

Comment:
With the removal of distinct "external" and "internal" endpoints and replacing them with a simple list, this has become a little complex. While I agree that there needs to be a more sophisticated solution to connect to a proper endpoint, I think that the complexity of the added logic may be a bit high for this refactoring chore. Do you want me to file a separate ticket to reimplement the endpoint choosing method in the new Roster world?

Reply:
Unfortunately, I think this has to be fixed in this PR. 

Here is a [mainnet.config.txt](https://github.com/swirlds/infrastructure/blob/cdcc7800538ec804858dbb96c0e4cfc3dc4ff46b/mainnet/ansible/config.mainnet.txt)

Multiple nodes list `10.x.x.x` IP addresses as internal addresses.   We will never successfully connect to these IP addresses if we attempt to connect to them.  We have to connect to the IP address in the external position.   If we don't fail over from attempting the first IP address, then we just fail and that isn't going to work.

Reply:
Yes, I considered that.

All the `10.x.x.x` IP addresses are listed as the first addresses in config.txt, and the `AddressBookUtils.parseAddressText` correctly uses the first address as an internal address, and then uses the second address as an external one.

The `RosterRetriever.buildRoster` puts the external address into a ServiceEndpoint with index 0, and the internal address into the ServiceEndpoint with index 1.

The refactored code basically defaults to using the ServiceEndpoint with index 0 at all times - that is whenever it opens a server socket, or a client socket to connect to a peer.

What all this means is that the refactored code will only ever use the external IP of a node to connect to it. Unless a sub-network hosting several nodes at once has a firewall setup in such a way that it prohibits the members of the sub-net to connect to their external IP addresses, this should work at all times and allow all nodes to be able to connect to all the other nodes. I don't believe we have firewalls configured in such a way, but even if we do, resolving this would be a matter of reconfiguring the firewall to allow such connections.

Like a comment above in this same class states, it would be ideal to have a more sophisticated logic in order to allow the use of local ip addresses when possible. But given that the Roster doesn't explicitly differentiate between local and external addresses, building such a logic may be not trivial.

Long story short, per the above reasoning, the refactoring code should work. If it needs to be improved, this could probably be done in a separate ticket, perhaps even unrelated to the TSS/Roster specifically, since, as mentioned in the port number-related conversation above, it may need even a bigger overhaul.

Reply:
I am asking in the DevOps channel if they or Solo  is expecting to make connections through the `internal` ip address.   If not, then creating an issue to perform the correct behavior and implementing it later is fine.  But if this PR breaks behavior, then we need to implement it as part of this PR.

Reply:
I've created an issue for this item and added it to the sprint.   https://github.com/hashgraph/hedera-services/issues/16065